message_sender/tasks.py

```diff
@@ -77,6 +80,71 @@ def run(self, metric_name, metric_value, session=None, **kwargs):
 fire_metric = FireMetric()
 
 
+class ConcurrencyLimiter(object):
+    BUCKET_SIZE = 60
+
+    @classmethod
+    def get_current_message_count(cls, msg_type, delay):
+        # Sum the values in all the buckets to get the total
+        total = 0
+        number_of_buckets = delay // cls.BUCKET_SIZE + 1
+        bucket = int(time.time() // cls.BUCKET_SIZE)
+        for i in range(0, number_of_buckets):
+            value = cache.get(msg_type+"_messages_at_%s" % bucket)
+            bucket = bucket - 1
+            if value:
+                total += int(value)
+        return total
+
+    @classmethod
+    def incr_message_count(cls, msg_type, delay):
+        bucket = int(time.time() // cls.BUCKET_SIZE)
+        key = msg_type + "_messages_at_%s" % bucket
+        value = cache.get(key)
```

Comment:
There's a possible race condition here. If two workers do the `get` before one of the workers do the `set`, then both workers will attempt to set the cache value to 1 (instead of one settings, and one incrementing).

I think using `cache.add` should solve this, something like:
```
if not cache.add(key, 1, delay + cls.BUCKET_SIZE):
    cache.incr(key)
```